File: claude-experiments/js-ir/crates/jsir-ssa/oracle/bench.js
#!/usr/bin/env node
// Warm throughput benchmark for the official React compiler: compile every
// fixture in ONE process (excludes Node startup), report ms/compile. Pair with
// `cargo run -p jsir-ssa --example bench` for a head-to-head. Set ITERS to vary.
const fs = require('fs');
const path = require('path');
const babel = require('@babel/core');
const RC = require('babel-plugin-react-compiler');
const plugin = RC.default;
const parse = RC.parseConfigPragmaForTests;

const dir = process.env.REACT_FIXTURES || path.join(__dirname, 'fixtures');
const files = fs.readdirSync(dir).filter((f) => f.endsWith('.js'));
const srcs = files.map((f) => fs.readFileSync(path.join(dir, f), 'utf8'));

function compile(src) {
  const nl = src.indexOf('\n');
  const first = nl < 0 ? src : src.slice(0, nl);
  const lang = first.includes('@flow') ? 'flow' : 'typescript';
  const cfg = parse(first, { compilationMode: 'all' });
  try {
    babel.transformSync(src, {
      filename: '/t.tsx',
      parserOpts: { plugins: lang === 'flow' ? ['flow', 'jsx'] : ['typescript', 'jsx'] },
      plugins: [[plugin, { ...cfg, environment: { ...cfg.environment } }]],
      sourceType: 'module',
      compact: false,
      configFile: false,
      babelrc: false,
    });
    return true;
  } catch (e) {
    return false;
  }
}

const iters = parseInt(process.env.ITERS || '10', 10);
const okset = []; let err = 0;
for (const s of srcs) (compile(s) ? okset.push(s) : err++); // warmup + partition
for (const [label, set] of [['ALL', srcs], ['OK-only', okset]]) {
  const t = process.hrtime.bigint();
  for (let i = 0; i < iters; i++) for (const s of set) compile(s);
  const el = Number(process.hrtime.bigint() - t) / 1e6;
  const total = set.length * iters;
  console.log(
    `REACT[${label}]: n=${set.length} ${el.toFixed(1)}ms => ${(el / total).toFixed(4)} ms/compile, ${(total / (el / 1000)).toFixed(0)} compiles/sec`,
  );
}
console.log(`REACT: fixtures=${srcs.length} ok=${okset.length} err=${err}`);

File: claude-experiments/js-ir/crates/jsir-ssa/examples/bench.rs
//! Throughput benchmark for our compiler: compile every fixture in-process,
//! warm, and report ms/compile. Pair with `oracle/bench.js` (the official React
//! compiler) for a head-to-head. Set REACT_FIXTURES / ITERS to override.
use std::time::Instant;

fn main() {
    let dir = std::env::var("REACT_FIXTURES")
        .unwrap_or_else(|_| "crates/jsir-ssa/oracle/fixtures".into());
    let mut srcs = Vec::new();
    for e in std::fs::read_dir(&dir).unwrap() {
        let p = e.unwrap().path();
        if p.extension().map(|x| x == "js").unwrap_or(false) {
            if let Ok(s) = std::fs::read_to_string(&p) {
                srcs.push(s);
            }
        }
    }
    let iters: usize = std::env::var("ITERS").ok().and_then(|s| s.parse().ok()).unwrap_or(10);

    // Warmup; keep only the fixtures we actually compile (exclude fast bails so
    // the per-compile figure reflects real work, not error paths).
    let mut okset: Vec<&String> = Vec::new();
    let mut err = 0usize;
    for s in &srcs {
        match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| jsir_ssa::codegen::compile(s))) {
            Ok(Ok(_)) => okset.push(s),
            _ => err += 1,
        }
    }

    for (label, set) in [("ALL", srcs.iter().collect::<Vec<_>>()), ("OK-only", okset.clone())] {
        let start = Instant::now();
        for _ in 0..iters {
            for s in &set {
                let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                    jsir_ssa::codegen::compile(s)
                }));
            }
        }
        let el = start.elapsed();
        let total = set.len() * iters;
        println!(
            "OURS[{label}]: n={} {:.1}ms => {:.4} ms/compile, {:.0} compiles/sec",
            set.len(),
            el.as_secs_f64() * 1000.0,
            el.as_secs_f64() * 1000.0 / total as f64,
            total as f64 / el.as_secs_f64()
        );
    }
    println!("OURS: fixtures={} ok={} err={}", srcs.len(), okset.len(), err);
}
